### PR TITLE
envoy.base.checker: Cleanups and features

### LINKED ISSUE #159: Improve `envoy.base.checker` summary
the checker summary could do with a bit of UX attention

also, there seems to be an issue where it displays summary notifications for a count of warnings/errors matching the least amount per check (or something like that)

https://dev.azure.com/cncf/envoy/_build/results?buildId=97768&view=logs&j=1c04237c-5f52-5959-5f83-5522b3556331&t=db3f3889-c922-566b-fa2e-4bc4948bc467&l=568

### PR DESCRIPTION
**Pull request by phlax**
cleanups:

- fix/improve summary (Fix #159)
- test cleanups

features:

- async check queue
- data preloading
- check disabling

Signed-off-by: Ryan Northey <ryan@synca.io>

### CODE REVIEW

File: envoy.base.checker/envoy/base/checker/checker.py [hunk 1]
```diff
@@ -371,20 +421,93 @@ def _section(self, message: str, lines: list = None) -> list:
...
+        if self._check_should_preload(check, handler):
+            return self.preload_task(check, handler["run"](self))
+
+    async def preload_task(self, name: str, task: Awaitable) -> None:
+        """Wrap a preload task with the pending queue, and trigger `on_preload`
+        event on completion."""
+        start = time.time()
+        self.log.debug(f"Preloading {name}")
+        self.preload_pending.append(name)
+        await task
```
Comment (phlax) [inline envoy.base.checker/envoy/base/checker/checker.py:507]: try/finally - place in removed if failed ?

File: envoy.base.checker/envoy/base/checker/checker.py [hunk 2]
```diff
@@ -371,20 +421,93 @@ def _section(self, message: str, lines: list = None) -> list:
...
+        # Place all checks that are not blocked in the queue.
+        for check in self.checks_to_run:
+            if check not in self.preload_checks:
+                await self.check_queue.put(check)
+
+    async def on_preload(self) -> None:
+        """Event fired after each preload task completes."""
+        for check in self.checks_to_run:
+            if self._check_should_run(check):
+                self.preloaded.append(check)
```
Comment (phlax) [inline envoy.base.checker/envoy/base/checker/checker.py:481]: move to preload_task

File: envoy.base.checker/envoy/base/checker/checker.py [hunk 3]
```diff
@@ -371,20 +421,93 @@ def _section(self, message: str, lines: list = None) -> list:
...
+                await self.check_queue.put(check)
+
+    def preload(self) -> Awaitable:
+        """Returns an awaitable of gathered preload tasks."""
+
+        async def _gather():
```
Comment (phlax) [inline envoy.base.checker/envoy/base/checker/checker.py:487]: make `preload` async and remove this

File: envoy.base.checker/envoy/base/checker/checker.py [hunk 4]
```diff
@@ -371,20 +421,93 @@ def _section(self, message: str, lines: list = None) -> list:
...
+            await asyncio.gather(*[t for t in self.preload_tasks])
+        return _gather()
+
+    def preload_data(
+            self,
+            check: str,
+            handler: Dict[str, Any]) -> Optional[Awaitable]:
+        """Return an awaitable preload task if required."""
+        if self._check_should_preload(check, handler):
+            return self.preload_task(check, handler["run"](self))
```
Comment (phlax) [inline envoy.base.checker/envoy/base/checker/checker.py:499]: fix the decorator so no need to pass self

File: envoy.base.checker/envoy/base/checker/checker.py [hunk 5]
```diff
@@ -325,26 +360,36 @@ def max_warnings(self) -> int:
...
-            _out.extend(self._section(msg, problems[:_max]))
-        if not _out:
-            return
+            self.print_failed_check(check, problems, problem_type)
+
+    def print_failed_check(self, check, problems, problem_type):
+        global_max = getattr(self, f"max_{problem_type}")
+        _out = []
+        _msg = ""
+        _max = (
```
Comment (phlax) [inline envoy.base.checker/envoy/base/checker/checker.py:370]: move to a method

File: envoy.base.checker/envoy/base/checker/checker.py [hunk 6]
```diff
@@ -325,26 +360,36 @@ def max_warnings(self) -> int:
...
+            min(len(problems), global_max)
+            if global_max >= 0
+            else len(problems))
+        msg = (
```
Comment (phlax) [inline envoy.base.checker/envoy/base/checker/checker.py:374]: ditto

File: envoy.base.checker/envoy/base/checker/checker.py [hunk 7]
```diff
@@ -371,20 +439,144 @@ def _section(self, message: str, lines: list = None) -> list:
...
+        """Tuple of awaitables for preloading check data."""
+        tasks = [
+            self.preload_data(name, handler)
+            for name, handler
+            in self.checks_data.items()]
+        return tuple(t for t in tasks if t)
+
+    @cached_property
+    def preloaded(self) -> List[str]:
+        """Completed preload tasks."""
```
Comment (phlax) [inline envoy.base.checker/envoy/base/checker/checker.py:477]: checks ?

File: envoy.base.checker/envoy/base/checker/decorators.py [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+        self.kwargs = kwargs
+
+    def __call__(self, fun, *args, **kwargs):
+        self._fun = fun
+        return self
+
+    def __set_name__(self, owner, name):
+        self.name = name
+        self.owner = owner
+        owner.checks_data = getattr(owner, "checks_data", {})
```
Comment (phlax) [inline envoy.base.checker/envoy/base/checker/decorators.py:18]: as this is writing on the class it needs to be a mutable data type

**Review by kfaseela — APPROVED** [review]

**Pull request merged**